src/main/scala/vectorpipe/sources/ChangeMicroBatchReader.scala

```diff
@@ -24,7 +24,7 @@ class ChangeStreamBatchReader(baseURI: URI, sequences: Seq[Int])
 
 case class ChangeMicroBatchReader(options: DataSourceOptions, checkpointLocation: String)
     extends ReplicationStreamMicroBatchReader[Change](options, checkpointLocation) {
-  private val baseURI = new URI(
+  private lazy val baseURI = new URI(
```

Comment:
I don't know exactly why this fixes the NullPointerException being thrown here. I do know that when baseURI is initially called (without the `lazy`),`options.get(Source.BaseURI)` returns `Optional<null>`.

Reply:
Any record of the NPE to share?

Reply:
```
[info] 19/12/16 15:55:47 ERROR sources.ChangesetSource: Error fetching / parsing changeset state.
[info] java.lang.NullPointerException
[info]  at vectorpipe.sources.ChangesetSource$.vectorpipe$sources$ChangesetSource$$callRealBody$1(ChangesetSource.scala:76)
[info]  at vectorpipe.sources.ChangesetSource$$anonfun$getCurrentSequence$1.apply(ChangesetSource.scala:72)
[info]  at vectorpipe.sources.ChangesetSource$$anonfun$getCurrentSequence$1.apply(ChangesetSource.scala:72)
[info]  at com.softwaremill.macmemo.MemoCacheBuilder$$anon$1$$anon$2$$anon$3.call(MemoCacheBuilder.scala:64)
[info]  at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4876)
[info]  at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
[info]  at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2277)
[info]  at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
[info]  at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
[info]  at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
[info]  at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871)
[info]  at com.softwaremill.macmemo.MemoCacheBuilder$$anon$1$$anon$2.get(MemoCacheBuilder.scala:63)
[info]  at vectorpipe.sources.ChangesetSource$.getCurrentSequence(ChangesetSource.scala:72)
[info]  at vectorpipe.sources.ChangesetMicroBatchReader.getCurrentSequence(ChangesetMicroBatchReader.scala:35)
[info]  at vectorpipe.sources.ReplicationStreamMicroBatchReader$$anonfun$5$$anonfun$apply$1.apply(ReplicationStreamMicroBatchReader.scala:4
6)
[info]  at vectorpipe.sources.ReplicationStreamMicroBatchReader$$anonfun$5$$anonfun$apply$1.apply(ReplicationStreamMicroBatchReader.scala:4
6)
[info]  at scala.Option.orElse(Option.scala:289)
[info]  at vectorpipe.sources.ReplicationStreamMicroBatchReader$$anonfun$5.apply(ReplicationStreamMicroBatchReader.scala:46)
[info]  at vectorpipe.sources.ReplicationStreamMicroBatchReader$$anonfun$5.apply(ReplicationStreamMicroBatchReader.scala:46)
[info]  at scala.Option.orElse(Option.scala:289)
[info]  at vectorpipe.sources.ReplicationStreamMicroBatchReader.<init>(ReplicationStreamMicroBatchReader.scala:46)
[info]  at vectorpipe.sources.ChangesetMicroBatchReader.<init>(ChangesetMicroBatchReader.scala:27)
[info]  at vectorpipe.sources.ChangesetProvider.createMicroBatchReader(ChangesetProvider.scala:32)
[info]  at org.apache.spark.sql.streaming.DataStreamReader.load(DataStreamReader.scala:182)
[info]  at vectorpipe.examples.ChangesetStreamProcessor$$anonfun$$lessinit$greater$1.apply(ChangesetStreamProcessor.scala:69)
[info]  at vectorpipe.examples.ChangesetStreamProcessor$$anonfun$$lessinit$greater$1.apply(ChangesetStreamProcessor.scala:57)
[info]  at cats.SemigroupalArityFunctions$$anonfun$map4$1.apply(SemigroupalArityFunctions.scala:60)
[info]  at cats.SemigroupalArityFunctions$$anonfun$map4$1.apply(SemigroupalArityFunctions.scala:60)
[info]  at cats.data.Validated.ap(Validated.scala:182)
[info]  at cats.data.ValidatedApplicative.ap(Validated.scala:527)
[info]  at cats.data.ValidatedApplicative.ap(Validated.scala:520)
[info]  at cats.ComposedApply$$anonfun$ap$1$$anonfun$apply$1.apply(Composed.scala:33)
[info]  at cats.Monad$$anonfun$map$1.apply(Monad.scala:16)
[info]  at cats.instances.Function0Instances$$anon$1$$anonfun$flatMap$1.apply(function.scala:74)
[info]  at cats.instances.Function0Instances$$anon$1$$anonfun$flatMap$1.apply(function.scala:74)
[info]  at com.monovore.decline.Parser.com$monovore$decline$Parser$$evalResult(Parser.scala:27)
[info]  at com.monovore.decline.Parser.consumeAll(Parser.scala:95)
[info]  at com.monovore.decline.Parser.apply(Parser.scala:18)
[info]  at com.monovore.decline.Command.parse(opts.scala:17)
[info]  at com.monovore.decline.CommandApp.main(CommandApp.scala:48)
[info]  at vectorpipe.examples.ChangesetStreamProcessor.main(ChangesetStreamProcessor.scala)
```

Reply:
Mark that in the "weird" column.  Then pack it, ship it.

Reply:
Very well. I'm not particularly satisfied, but I expect I could spend hours on this rabbit hole, and presumably its "correct" anyway to only initialize the baseURI once at the time we need it and not before. Hopefully we're not doing the classic "paper over a effect without solving the root cause" here.